File: lib/apt/src/main/java/site/addzero/apt/DictEnumCodeGenerator.java
package site.addzero.apt;

import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.tools.Diagnostic;
import javax.tools.JavaFileObject;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.*;
import java.util.stream.Collectors;

/**
 * 字典枚举代码生成器 - 生成 Java 风格的枚举类
 */
public class DictEnumCodeGenerator {

    private final Filer filer;
    private final Messager messager;
    private final String enumOutputPackage;

    public DictEnumCodeGenerator(Filer filer, Messager messager, String enumOutputPackage) {
        this.filer = filer;
        this.messager = messager;
        this.enumOutputPackage = enumOutputPackage;
    }

    /**
     * 生成所有字典的枚举类
     */
    public void generateEnumClasses(List<DictMetadataExtractor.DictMetadata> dictMetadataList) {
        Set<String> generatedClassNames = new HashSet<>();

        dictMetadataList.forEach(dictMetadata -> {
            try {
                if (dictMetadata.items.isEmpty()) {
                    messager.printMessage(
                        Diagnostic.Kind.WARNING,
                        "字典'" + dictMetadata.dictCode + "'没有字典项,跳过枚举生成"
                    );
                    return;
                }

                String enumName = toCamelCase(dictMetadata.dictCode, true);

                if (generatedClassNames.contains(enumName)) {
                    messager.printMessage(
                        Diagnostic.Kind.WARNING,
                        "字典'" + dictMetadata.dictCode + "'生成的枚举类名'" + enumName + "'重复,跳过"
                    );
                    return;
                }

                generatedClassNames.add(enumName);
                generateEnumClass(dictMetadata, enumName);
            } catch (Exception e) {
                messager.printMessage(
                    Diagnostic.Kind.WARNING,
                    "为字典'" + dictMetadata.dictCode + "'生成枚举类失败: " + e.getMessage()
                );
            }
        });
    }

    /**
     * 生成单个枚举类
     */
    private void generateEnumClass(DictMetadataExtractor.DictMetadata dictMetadata, String enumName) {
        String fullEnumName = "Enum" + enumName;
        String qualifiedName = enumOutputPackage + "." + fullEnumName;

        try {
            JavaFileObject fileObject = filer.createSourceFile(qualifiedName);
            try (PrintWriter writer = new PrintWriter(fileObject.openWriter())) {
                writer.println("package " + enumOutputPackage + ";");
                writer.println();
                writer.println("/**");
                writer.println(" * " + dictMetadata.dictName);
                writer.println(" *");
                writer.println(" * 数据库字典编码: " + dictMetadata.dictCode);
                writer.println(" * 自动生成的枚举类，不要手动修改");
                writer.println(" */");
                writer.println("public enum " + fullEnumName + " {");
                writer.println();
                
                // 生成枚举常量
                List<String> enumConstants = dictMetadata.items.stream()
                    .map(item -> {
                        String enumEntryName = toEnumEntryName(item.desc, item.code);
                        return String.format("    %s(\"%s\", \"%s\")",
                            enumEntryName,
                            escapeJavaString(item.code),
                            escapeJavaString(item.desc));
                    })
                    .collect(Collectors.toList());
                
                writer.println(String.join(",\n", enumConstants) + ";");
                writer.println();
                
                // 生成字段
                writer.println("    private final String code;");
                writer.println("    private final String desc;");
                writer.println();
                
                // 生成构造函数
                writer.println("    " + fullEnumName + "(String code, String desc) {");
                writer.println("        this.code = code;");
                writer.println("        this.desc = desc;");
                writer.println("    }");
                writer.println();
                
                // 生成 getter 方法
                writer.println("    public String getCode() {");
                writer.println("        return code;");
                writer.println("    }");
                writer.println();
                writer.println("    public String getDesc() {");
                writer.println("        return desc;");
                writer.println("    }");
                writer.println();
                
                // 生成工具方法
                writer.println("    /**");
                writer.println("     * 根据编码获取枚举值");
                writer.println("     *");
                writer.println("     * @param code 编码");
                writer.println("     * @return 对应的枚举值，如果不存在则返回null");
                writer.println("     */");
                writer.println("    public static " + fullEnumName + " fromCode(String code) {");
                writer.println("        if (code == null) {");
                writer.println("            return null;");
                writer.println("        }");
                writer.println("        for (" + fullEnumName + " e : values()) {");
                writer.println("            if (e.code.equals(code)) {");
                writer.println("                return e;");
                writer.println("            }");
                writer.println("        }");
                writer.println("        return null;");
                writer.println("    }");
                writer.println();
                writer.println("    /**");
                writer.println("     * 根据描述获取枚举值");
                writer.println("     *");
                writer.println("     * @param desc 描述");
                writer.println("     * @return 对应的枚举值，如果不存在则返回null");
                writer.println("     */");
                writer.println("    public static " + fullEnumName + " fromDesc(String desc) {");
                writer.println("        if (desc == null) {");
                writer.println("            return null;");
                writer.println("        }");
                writer.println("        for (" + fullEnumName + " e : values()) {");
                writer.println("            if (e.desc.equals(desc)) {");
                writer.println("                return e;");
                writer.println("            }");
                writer.println("        }");
                writer.println("        return null;");
                writer.println("    }");
                
                writer.println("}");
            }
            
            messager.printMessage(Diagnostic.Kind.NOTE, "Generated enum class: " + fullEnumName);
        } catch (IOException e) {
            messager.printMessage(
                Diagnostic.Kind.ERROR,
                "Failed to create enum class " + fullEnumName + ": " + e.getMessage()
            );
        }
    }

    /**
     * 转换为驼峰命名
     */
    private String toCamelCase(String str, boolean capitalizeFirst) {
        String[] words = str.replaceAll("[_\\-\\s]+", " ")
                            .split(" ");
        
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            if (word.isEmpty()) continue;
            
            if (i == 0 && !capitalizeFirst) {
                result.append(word.toLowerCase());
            } else {
                result.append(Character.toUpperCase(word.charAt(0)));
                if (word.length() > 1) {
                    result.append(word.substring(1).toLowerCase());
                }
            }
        }
        
        return result.toString();
    }

    /**
     * 转换为合法的枚举项名称
     */
    private String toEnumEntryName(String desc, String code) {
        // 优先使用 code，如果 code 为空则使用 desc
        String name = (code != null && !code.isEmpty()) ? code : desc;
        if (name == null || name.isEmpty()) {
            name = "UNKNOWN";
        }
        
        // 转换为大写并替换非法字符
        name = name.toUpperCase()
                   .replaceAll("[^A-Z0-9_]", "_")
                   .replaceAll("_+", "_");
        
        // 确保以字母或下划线开头
        if (!name.isEmpty() && Character.isDigit(name.charAt(0))) {
            name = "_" + name;
        }
        
        // 移除首尾下划线
        name = name.replaceAll("^_+|_+$", "");
        
        if (name.isEmpty()) {
            name = "UNKNOWN";
        }
        
        return name;
    }

    /**
     * 转义 Java 字符串
     */
    private String escapeJavaString(String str) {
        if (str == null) {
            return "";
        }
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\n", "\\n")
                  .replace("\r", "\\r")
                  .replace("\t", "\\t");
    }
}

File: lib/ksp/jdbc2metadata/addzero-jdbc2enum-processor/src/commonMain/kotlin/site/addzero/DictEnumCodeGenerator.kt
package site.addzero

import site.addzero.context.SettingContext.settings
import site.addzero.util.PinYin4JUtils
import site.addzero.util.str.isNotBlank
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import java.io.File
import java.util.Locale.getDefault

/**
 * 字典枚举代码生成器
 *
 * 负责根据字典元数据生成枚举类代码
 */
class DictEnumCodeGenerator(
    private val codeGenerator: CodeGenerator,
    private val logger: KSPLogger,
) {

    /**
     * 生成所有字典的枚举类
     *
     * @param dictMetadataList 字典元数据列表
     */
    fun generateEnumClasses(dictMetadataList: List<DictMetadataExtractor.DictMetadata>) {
        // 生成的类名集合，用于检测重复
        val generatedClassNames = mutableSetOf<String>()

        dictMetadataList.forEach { dictMetadata ->
            try {
                // 检查是否有字典项
                if (dictMetadata.items.isEmpty()) {
                    logger.warn("字典'${dictMetadata.dictCode}'没有字典项,跳过枚举生成")
                    return@forEach
                }

                // 转换为类名
                val enumName = toCamelCase(dictMetadata.dictCode, true)

                // 检查类名是否重复
                if (enumName in generatedClassNames) {
                    logger.warn("字典'${dictMetadata.dictCode}'生成的枚举类名'$enumName'重复,跳过")
                    return@forEach
                }

                // 记录类名
                generatedClassNames.add(enumName)

                // 生成枚举类
                generateEnumClass(dictMetadata, enumName)
//                logger.info("为字典'${dictMetadata.dictCode}'生成枚举类: Enum$enumName")
            } catch (e: Exception) {
                logger.warn("为字典'${dictMetadata.dictCode}'生成枚举类失败: ${e.message}")
            }
        }
    }
//    /**
//     * 枚举输出目录（shared 编译目录）
//     */
//    val enumOutputDir: String = sharedSourceDir.withPkg(enumOutputPackage)
    /**
     * 生成单个枚举类
     *
     * @param dictMetadata 字典元数据
     * @param enumName 枚举类名（不含前缀）
     */
    private fun generateEnumClass(dictMetadata: DictMetadataExtractor.DictMetadata, enumName: String) {
        // 枚举类名加上"Enum"前缀
        val fullEnumName = "Enum$enumName"

        // 构建枚举类内容
        val enumOutputPackage = settings.enumOutputPackage
        val enumOutputDir = settings.enumOutputDir
        val enumContent = """
           package $enumOutputPackage
            /**
             * ${dictMetadata.dictName}
             * 
             * 数据库字典编码: ${dictMetadata.dictCode}
             * 自动生成的枚举类，不要手动修改
             */
            enum class $fullEnumName(
                val code: String,
                val desc: String
            ) {
                ${generateEnumEntries(dictMetadata.items)};
            }
        """.trimIndent()

        // 创建枚举类文件
        // 使用手动 File IO 生成到指定目录
        writeEnumToFile(fullEnumName, enumContent, enumOutputDir)
    }

    /**
     * 生成枚举项
     *
     * @param items 字典项列表
     * @return 枚举项字符串
     */
    private fun generateEnumEntries(items: List<DictMetadataExtractor.DictItem>): String {
        return items.joinToString(",\n    ") { item ->
            val enumEntryName = toEnumEntryName(item.desc, item.code)
            """$enumEntryName("${item.code}", "${item.desc}")"""
        }
    }

    /**
     * 转换为驼峰命名
     *
     * @param str 输入字符串
     * @param capitalizeFirst 是否将首字母大写
     * @return 驼峰命名字符串
     */
    private fun toCamelCase(str: String, capitalizeFirst: Boolean): String {
        // 将下划线、中划线等分隔符替换为空格
        val words = str.replace(Regex("[_\\-\\s]+"), " ")
            .split(" ")
            .filter { it.isNotEmpty() }

        if (words.isEmpty()) return ""

        return words.mapIndexed { index, word ->
            if (index == 0 && !capitalizeFirst) {
                word.lowercase()
            } else {
                word.replaceFirstChar { if (it.isLowerCase()) it.titlecase(getDefault()) else it.toString() }
            }
        }.joinToString("")
    }

    /**
     * 使用手动 File IO 写入枚举文件
     */
    private fun writeEnumToFile(fileName: String, content: String, outputDir: String) {


        val enumOutputDir = settings.enumOutputDir

        // 创建目录
        val dir = File(enumOutputDir)
        dir.mkdirs()

        // 写入文件
        val file = File(dir, "$fileName.kt")
        file.writeText(content)

        logger.info("Generated enum class: $fileName at ${file.absolutePath}")
    }

    /**
     * 转换为合法的枚举项名称
     *
     * @param desc 字典项描述
     * @param code 字典项编码，用于desc为空时的备选
     * @return 合法的枚举项名称
     */
    private fun toEnumEntryName(desc: String, code: String): String {
        // 如果描述为空，则直接使用code
        if (desc.isEmpty()) {
            return PinYin4JUtils.sanitize(code)
        }

        if (code.isNotBlank() && desc.isNotBlank()) {
            return PinYin4JUtils.sanitize(code)
        }

        try {
            // 1. 尝试将中文转为拼音
            val pinyin = PinYin4JUtils.hanziToPinyin(desc, "_")

            // 2. 如果拼音转换成功，将其处理为合法变量名并返回
            if (pinyin.isNotBlank()) {
                return PinYin4JUtils.sanitize(pinyin)
            }
        } catch (e: Exception) {
            logger.warn("Failed to convert '${desc}' to pinyin: ${e.message}")
        }

        // 3. 如果拼音转换失败，直接尝试将原始描述处理为合法变量名
        val sanitizedDesc = PinYin4JUtils.sanitize(desc)

        // 4. 如果处理后的描述不为空，返回它
        if (sanitizedDesc != "CODE") {
            return sanitizedDesc
        }

        // 5. 最后兜底：使用code生成变量名
        return PinYin4JUtils.sanitize(code)
    }
}

private fun genTools(fullEnumName: String): String = """                
    companion object {
                    /**
                     * 根据编码获取枚举值
                     * 
                     * @param code 编码
                     * @return 对应的枚举值，如果不存在则返回null
                     */
                    fun fromCode(code: String): $fullEnumName? {
                        return entries.find { it.code == code }
                    }
                    
                    /**
                     * 根据描述获取枚举值
                     * 
                     * @param desc 描述
                     * @return 对应的枚举值，如果不存在则返回null
                     */
                    fun fromDesc(desc: String): $fullEnumName? {
                        return entries.find { it.desc == desc }
                    }
                }
"""
